File: Algorithm(2022)/BFSAndDFS/DFS01.go
package main

import "fmt"

func dfs(graph [][]int, v int, visited []bool) {
	visited[v] = true
	fmt.Printf("%d ", v)
	for _, i := range graph[v] {
		if !(visited[i]) {
			dfs(graph, i, visited)
		}
	}
}

func main() {
	visited := [9]bool{}
	graph := [9][]int{
		{},
		{2, 3, 8},
		{1, 7},
		{1, 4, 5},
		{3, 5},
		{3, 4},
		{7},
		{2, 6, 8},
		{1, 7},
	}

	dfs(graph[:], 1, visited[:])
	fmt.Println()
}

File: Algorithm(2022)/BFSAndDFS/DFS01.scala
package BFSAndDFS

object DFS01 {
    def dfs(graph: Array[Array[Int]], v: Int, visited: Array[Boolean]): Unit = {
        visited(v) = true
        printf("%d ", v)
        for (i <- graph(v)) {
            if (!visited(i)) dfs(graph, i, visited)
        }
    }
    def main(args: Array[String]): Unit = {
        val graph = Array(
            Array(0),
            Array(2, 3, 8),
            Array(1, 7),
            Array(1, 4, 5),
            Array(3, 5),
            Array(3, 4),
            Array(7),
            Array(2, 6, 8),
            Array(1, 7)
        )
        var visited = Array.ofDim[Boolean](9)

        dfs(graph, 1, visited)
        println()
    }
}
